### ISSUE TITLE
Warn users on all incompatible mods

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#91]
Currently I think we only check two things:

* Is old TMPE installed?
* Has anything detectably messed with the harmony patches?

IMO this is insufficient, given that we have a list of known incompatible mods:

https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/wiki/Incompatible-mods

We should at least put messages in game log if any of those mods are subscribed (regardless of whether they are enabled).

Ideally we should put message on screen (regardless of that 'show error' setting in mod options) to let user know they have certifiably incompatible mods installed.

**Comment by krzychu124** [comment]
I can add other steamId's for checking while starting game but I will have to refactor that show popup message action because I am sure we don't want to spam user with popups 😄

**Comment by aubergine10** [comment]
I'll create text file and submit a PR - where do you want it going? ```/resources```?

**Comment by krzychu124** [comment]
I think ```/resources``` will be good place for it, thanks.

**Comment by aubergine10** [comment]
Do you need link to workshop? It can be constructed:

```
https://steamcommunity.com/sharedfiles/filedetails/?id=%steamid%
```

?

**Comment by krzychu124** [comment]
Maybe not... I forgot about that 😃

**Comment by aubergine10** [comment]
I also found a few additional incompatible mods so the wiki page was also updated :)

**Comment by aubergine10** [comment]
Checked over entire remaining list to make sure no more blunders. All present and correct :)

Over to you for coding magic :)

**Comment by aubergine10** [comment]
Just built a Steam "Collection" of all the incompatible mods - needs some images adding and stuff but should be able to get that done tomorrow. Users will be able to unsubscribe all of them with single click! Will post link when it's ready.

**Comment by krzychu124** [comment]
I will try to check how to call Steam workshop browser using code

**Comment by pcfantasy** [comment]
@aubergine10 

Advanced Traffic Congestion Report is not replaced by TMPE.

I add a pull to fix miss noiseDensity logic in TMPE, but Advanced Traffic Congestion Report will use another method to caculate trafficDensity and give to vannila Traffic Congestion Report map to show.

The two mods are either not confict and not overlap.

**Comment by krzychu124** [comment]
@pcfantasy good catch, we will fix that, thank you :wink:

**Comment by aubergine10** [comment]
Would it be possible to bring that feature in to TMPE?

The distinction between traffic **density** and traffic **congestion**.

Currently the TMPE dynamic lane selection seems based on density, not congestion. I think DLS would work much better if it operated on congestion. And path finding in general.

Also I think it would be extremely useful if the traffic info view had two tabs; one for density and one for congestion.

When I build new roads both factors should play a role in my decision making.

For example, if I want traffic moving faster, I need to be looking at the congestion and finding the bottlenecks. Removing choke points will often have huge impact on traffic density.

Currently all I can see in game is the density info view. But that is also useful, as it shows more where people are travelling, so I know if I change a popular route I should expect some consequences.

Anyway, I will remove the Advanced Traffic Congestion Report from the various lists (txt file in main repo, wiki page, and steam collection) of incompatible mods.

But if we could get that congestion feature in TMPE it would be hugely useful I think. Also, congestion could in the future lead to increased air pollution, contrasting to density which leads to increased noise pollution.

**Comment by krzychu124** [comment]
> Currently the TMPE dynamic lane selection seems based on density, not congestion. I think DLS would work much better if it operated on congestion. And path finding in general.

First I have to understand how that pathfinding sorcery works 😄 
> Also I think it would be extremely useful if the traffic info view had two tabs; one for density and one for congestion. 

Two tabs should be possible and almost everything related with GUI (time consuming research)

**Comment by aubergine10** [comment]
Here's a steam collection of the incompatible mods - please update mod description to include link to this (it's much easier for users to see what incompat mods they have and unsub them):

https://steamcommunity.com/sharedfiles/filedetails/?id=1658151346

**Comment by krzychu124** [comment]
Great, I will update steam workshop page today.

BTW I created quick incompatible mods tester.
I looked through code with dnSpy and I can see there is method to unsubscribe workśhop item using steamId.

Additionally I will play with dialog window if I find some time to directly render that list and unsubscribe them using button.
Would be great to open your list directly from game or fetch it using code(I've seen some methods, didn't test yet). You would update list with one click :wink:

**Comment by aubergine10** [comment]
Apparently ```Geli-Districts 3.0``` is also incompatible https://steamcommunity.com/sharedfiles/filedetails/?id=927293560

I've asked on their  workshop page to get in touch and let us know what the issue is.

**Comment by krzychu124** [comment]
Current progress...

![zrzut ekranu 112](https://user-images.githubusercontent.com/19638970/53290048-1e593000-379f-11e9-993c-a069bf477a60.png)

Issues:
- scrollbar not working
- mousewheel working only ina area of scrollbar track,
- I couldn't properly remove from modal stack (if opened from other modal, blur effect was still visible after closing if used
 ```
UiView.pushModal(IncompatibleModsPanel.instance); 
UiView.PopModal() - closes last opened modal
```
- in case now it is added as regular uicomponent attached to view you can click items on backdrop
- unsubscribed item is not removed from list (but I'm hiding button so not big issue)

**Comment by krzychu124** [comment]
I think I will switch my implementation to one big button with text __"Unsubscribe!"__ which will open steam workshop overlay with address to your collection of incompatible mods 😄

**Comment by aubergine10** [comment]
To prevent clicks and key presses filtering through to background, I think you have to do ```p,Use()``` on the event handlers.

For example:

```
        protected override void OnMouseDown(UIMouseEventParameter p)
        {
            if (p.buttons.IsFlagSet(UIMouseButton.Right))
            {
                Debug.Log("TMPEBB - Chameleon button - start drag");
                m_dragging = true;
                p.Use(); // mark the event as used and it should be filtered from parents/background
            }
            else
            {
                base.OnMouseDown(p);
            }
        }
```

**Comment by aubergine10** [comment]
BTW, if you change to a button that links to steam workshop, be careful - most if not all Mac users have the Steam Overlay turned off (it does weird stuff to the main menu screen and makes game unplayable).

**Comment by krzychu124** [comment]
I know, but there is other problem, because if it's opened from modal it's visible (because of BringToFront()) but you can't click anything - it's transparent to events..
For example when I open it by enabling mod from Content Manager (mod check trigger on enable), it became visible on top of Content Manager but that manager dialog blocks clicks - has higher priority 😕

**Comment by krzychu124** [comment]
I have managed to fix scrolling issue 😄

**Comment by aubergine10** [comment]
Fixed by #106
